### PR TITLE
Group all the logged gradients under the same sub-folder

### LINKED ISSUE #7711: Neptune logs folder is flooded when using track_grad_norm
## 🐛 Bug
If you try to log the norms of your gradients during training, they will be saved in the same folder as all the other logs, flooding that folder with a lot of attributes if you are training a model with a lot of weights

### To Reproduce

```
neptune_logger = NeptuneLogger(...)
trainer = pl.Trainer(logger=neptune_logger, track_grad_norm=2, ...)
```

### Expected behavior

I'm not sure what is the best solution to this problem, but I would suggest to put these logs in another folder inside "logs", like "logs/grad_norms" (it can be done as shown [here](https://docs.neptune.ai/you-should-know/logging-metadata) under "namespaces")

### Environment

 - PyTorch Version: 1.2.10
 - OS: Linux

### PR DESCRIPTION
**Pull request by alessiobonfiglio**
## What does this PR do?
Group all the tracked gradients norms to a subfolder instead of flooding the logs one
(NOTE: this has been tested only on neptune.ai, not with other loggers)
Fixes #7711

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
@alessiobonfiglio I think this does not generalise to all loggers, since not all of them have '/' as a separator (or at least this is not safe to assume). Maybe introduce an argument/logger property for this?

**Comment by alessiobonfiglio** [comment]
So should I add a method like `get_subfolder_separator()` to `LightningLoggerBase` that returns an attribute that is set to "_" and change it in the neptune logger (and any others you are going to suggest) constructor to "/"? Any suggestion for the names?

**Comment by justusschock** [comment]
@alessiobonfiglio maybe just add the following to the logger base:

```python
@property
def folder_separator(self):
    return '/'
```

I'll investigate tomorrow, which of them has which separator. Just want to make sure, this is future proof :)

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -457,7 +457,7 @@ def _track_gradient_norm(self):
         if (self.global_step + 1) % self.trainer.log_every_n_steps == 0:
             if float(self.trainer.track_grad_norm) > 0:
                 model = self.trainer.lightning_module
-                grad_norm_dict = grad_norm(model, self.trainer.track_grad_norm)
+                grad_norm_dict = grad_norm(model, self.trainer.track_grad_norm, self.trainer.folder_separator)
```
Comment (justusschock) [inline pytorch_lightning/trainer/training_loop.py:460]: There is the access to the logger missing.

Comment (alessiobonfiglio) [reply pytorch_lightning/trainer/training_loop.py]: Fixed, sorry I was in a hurry and I didn't proof check

File: pytorch_lightning/loggers/base.py [hunk 2]
```diff
@@ -335,6 +335,13 @@ def save_dir(self) -> Optional[str]:
         """
         return None
 
+    @property
+    def folder_separator(self):
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:339]: I propose the name group_separator or grouping separator to distinguish it from a filesystem folder structure.

File: pytorch_lightning/utilities/grads.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 from torch.nn import Module
 
 
-def grad_norm(module: Module, norm_type: Union[float, int, str]) -> Dict[str, float]:
+def grad_norm(module: Module, norm_type: Union[float, int, str], folder_separator: str) -> Dict[str, float]:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/grads.py:23]:
```suggestion
def grad_norm(module: Module, norm_type: Union[float, int, str], folder_separator: str = '/') -> Dict[str, float]:
```
Let's define a default here so we don't force changing code for all callers

**Comment by alessiobonfiglio** [comment]
And what should I test precisely? I mean, the reason behind this pr is #7711, but I don't think that it is testable

**Comment by alessiobonfiglio** [comment]
Any news? Is this good?

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by tchaton** [comment]
Dear @alessiobonfiglio,

Mind resolving the conflicts ? And the PR should be good to go :)

Best,
T.C

File: pytorch_lightning/loggers/base.py [hunk 4]
```diff
@@ -335,6 +335,13 @@ def save_dir(self) -> Optional[str]:
         """
         return None
 
+    @property
+    def group_separator(self):
+        """
+        Return the default separator used by the logger to group the data into subfolders.
+        """
+        return '/'
```
Comment (Borda) [inline pytorch_lightning/loggers/base.py:343]: it this also used for the dir/file path?

Comment (alessiobonfiglio) [reply pytorch_lightning/loggers/base.py:343]: Right now it is used only for put the logs of the norms of the gradients in their specific subfolder. I'm not sure to which files or directory you are referring, but I think this can be also used to create more complex hierarchies also when storing file artifacts with a logger

Comment (awaelchli) [reply pytorch_lightning/loggers/base.py:343]: this for grouping logs, has nothing to do with file paths. 
the separator for file parts should be determined by the filesystem type and operating system.

**Review by Borda — APPROVED** [review]

**Comment by kamil-kaczmarek** [comment]
@Borda @awaelchli, @tchaton 

What's the status of this PR?
It's great improvement of the gradients logging, worth introducing in the future release.

**Comment by alessiobonfiglio** [comment]
Done
